### PR TITLE
[Merged by Bors] - Add golang stat hooks

### LINKED ISSUE #856: Support stat operation for Golang on Linux
Related to this comment https://github.com/metalbear-co/mirrord/issues/789#issuecomment-1361130547

Some of our services are doing `stat` to do some check before opening the file, so it would be great if mirrord can support it.

Not a major blocker, feel free to treat this as low priority :)

### PR DESCRIPTION
**Pull request by infiniteregrets**
closes #856

### CODE REVIEW

File: mirrord/layer/src/go_hooks.rs [hunk 1]
```diff
@@ -406,18 +406,28 @@ unsafe extern "C" fn c_abi_syscall6_handler(
...
+                // |-- fstatat(_AT_FDCWD, path, stat, 0)
+                // - SYS_fstat will use fstat_detour, maps to the same syscall number i.e. SYS_FSTAT
+                //   (5)
+                // - SYS_newfstatat will use fstatat_detour, maps to the same syscall number i.e.
+                //   SYS_NEWFSTATAT (262)
+                // - SYS_lstat: maps to fstatat with AT_FDCWD and AT_SYMLINK_NOFOLLOW in go - no
+                //   additional hook needed
+                // - SYS_statx: not supported in go
+                libc::SYS_newfstatat => {
+                    match fstatat_logic(param1 as _, param2 as _, param3 as _, param4 as _) {
```
Comment (aviramha) [inline mirrord/layer/src/go_hooks.rs:420]: use `bypass_with` instead of match

File: mirrord/layer/src/go_hooks.rs [hunk 2]
```diff
@@ -406,18 +407,30 @@ unsafe extern "C" fn c_abi_syscall6_handler(
...
+                // |-- fstatat(_AT_FDCWD, path, stat, 0)
+                // - SYS_fstat will use fstat_detour, maps to the same syscall number i.e. SYS_FSTAT
+                //   (5)
+                // - SYS_newfstatat will use fstatat_detour, maps to the same syscall number i.e.
+                //   SYS_NEWFSTATAT (262)
+                // - SYS_lstat: maps to fstatat with AT_FDCWD and AT_SYMLINK_NOFOLLOW in go - no
+                //   additional hook needed
+                // - SYS_statx: not supported in go
+                libc::SYS_newfstatat => {
+                    let (Ok(result) | Err(result)) =
```
Comment (aviramha) [inline mirrord/layer/src/go_hooks.rs:421]: do we need the let here?

File: mirrord/layer/src/file/hooks.rs [hunk 3]
```diff
@@ -500,22 +500,31 @@ pub(crate) unsafe extern "C" fn __xstat_detour(
...
+        let res = res.metadata;
+        fill_stat(out_stat, &res);
+        0
+    })
+}
+
 /// Hook for `libc::fstatat`.
 #[hook_guard_fn]
-unsafe extern "C" fn fstatat_detour(
+pub(crate) unsafe extern "C" fn fstatat_detour(
```
Comment (aviramha) [inline mirrord/layer/src/file/hooks.rs:521]: no need to be pub now

File: mirrord/layer/src/file/hooks.rs [hunk 4]
```diff
@@ -500,22 +500,31 @@ pub(crate) unsafe extern "C" fn __xstat_detour(
...
 ) -> c_int {
-    let follow_symlink = (flag & libc::AT_SYMLINK_NOFOLLOW) == 0;
-    let path = (!raw_path.is_null()).then(|| CStr::from_ptr(raw_path));
-    let (Ok(result) | Err(result)) = xstat(Some(path), Some(fd), follow_symlink)
-        .map(|res| {
-            let res = res.metadata;
-            fill_stat(out_stat, &res);
-            0
-        })
+    let (Ok(result) | Err(result)) = fstatat_logic(fd, raw_path, out_stat, flag)
```
Comment (aviramha) [inline mirrord/layer/src/file/hooks.rs:527]: same here, I think we don't need let

**Review by aviramha — COMMENTED** [review]
Almost there! Great job

File: mirrord/layer/src/go_hooks.rs (hunk 2 above)
Comment (infiniteregrets) [reply mirrord/layer/src/go_hooks.rs:421]: do you mean to say we do `unwrap_or_else`?

**Review by aviramha — APPROVED** [review]

**Pull request closed without merging**